docs/proposals/env-vars.md

```diff
@@ -0,0 +1,563 @@
+	// Values defined by an Env with a duplicate key will take precedence.
+	// Cannot be updated.
+	// +optional
+	EnvFrom []EnvFromSource `json:"envFrom,omitempty" protobuf:"bytes,19,rep,name=envFrom"`
+	// List of environment variables to set in the container.
+	// Cannot be updated.
+	// +optional
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Env []EnvVar `json:"env,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,7,rep,name=env"`
+```
+
+First decision:  do we need to expose both `EnvFrom` and `EnvVar` off of `Containers`, especially given that
+`EnvVar` has an optional `EnvFromSource` [ref](https://github.com/kubernetes/api/blob/v0.21.0/core/v1/types.go#L1929)
+
+```go
+// EnvVar represents an environment variable present in a Container.
+type EnvVar struct {
+	// Name of the environment variable. Must be a C_IDENTIFIER.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// Optional: no more than one of the following may be specified.
+
+	// Variable references $(VAR_NAME) are expanded
+	// using the previous defined environment variables in the container and
+	// any service environment variables. If a variable cannot be resolved,
+	// the reference in the input string will be unchanged. The $(VAR_NAME)
+	// syntax can be escaped with a double $$, ie: $$(VAR_NAME). Escaped
+	// references will never be expanded, regardless of whether the variable
+	// exists or not.
+	// Defaults to "".
+	// +optional
+	Value string `json:"value,omitempty" protobuf:"bytes,2,opt,name=value"`
+	// Source for the environment variable's value. Cannot be used if value is not empty.
+	// +optional
+	ValueFrom *EnvVarSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
+```
+
+This proposal is making the simplifying assumption that ultimately setting the `Env []EnvVar` field in the 
+underlying Tekton `Step`/`Container` is sufficient for an initial implementation and scenarios we want to support.
+That will help the initial implementation in that the explicit "name" index field helps with default kubernetes 
+"path" API operations that tend to occur with devops/gitops flows.  Also, `ValueFrom *EnvVarSource` still allows
+us to pull content from `Secrets`, like authentication token.  Many of the tools when building images (including
+source code extraction and tools that have to download dependencies) consume authentication tokens via environment
+variables.  Storing those token in k8s `Secrets` vs. having to specify them in scripts referenced in `Containers` has 
+become a preferred method when building images in k8s native environments.
+
+Next, the question of whether to use the k8s `EnvVar` directly exists.  This proposal suggests creating a new
+Shipwright type that inlines the k8s type, in case we want to extend and add new features later that require
+additional fields or new types from k8s or Tekton.  So with that, we start with
+
+```go
+type EnvironmentVariable struct {
```

Comment:
Shorten to `EnvVar` (and `envVar` in our YAML representation?)

Reply:
I was worried about confusion with the existing k8s structs, but I'll acquiesce, unless others jump in with that similar concern.  I'll give a day or so to see if that occurs before making the switch.

Reply:
We should use k8s namings if they correspond directly to what gets added to a pod spec. Developers won't need to context switch.